# Article 9


## Objectif:

Cartographier les acteurs de la proposition de loi de l'article 9

### Découverte du dataset

Un fichier json avec les donnees relatives a l'article 9 est disponible
article9.json

    ! Le détail des votes par utilisateurs n'est disponible que dans le fichier participants

Ce fichier représente les données sous forme de dictionnaire: accessible par clé,valeur nous l'avons 
stocké dans la variable **article9**

La première démarche pour connaîre les données et d'afficher les clés disponibles du dictionnaires

Affichons les clés du dictionnaire:

In [1]:
import json, pprint
#afficher les données plus zouliement avec indentation
pp = pprint.PrettyPrinter(indent=4, depth=4)
with open("article9.json", "r") as f:
    article9 = json.load(f)
    pp.pprint(article9.keys())

[   u'subtitle',
    u'cat_id',
    u'updated_at',
    u'sources',
    u'arguments_count',
    u'versions_count',
    u'votes_total',
    u'author',
    u'article_link',
    u'arguments',
    u'answer',
    u'body',
    u'ranking',
    u'sources_count',
    u'votes_ok',
    u'versions',
    u'article_id',
    u'body_anchors',
    u'created_at',
    u'title',
    u'votes_mitige',
    u'body_links',
    u'votes_nok']


* Statistiques descriptives de l'article 9:

Nous allons charger les informations disponibles dans une **variable** appelée re_article9
qui va contenir les differents champs statistiques disponibles en les réorganisant légèrement pour une plus grande facilité de manipulation. 
Ici nous filtrons les clés du dictionnaires en fonction de leur nom que nous connaissons grace à la manipulation 
ci-dessus.

In [ ]:
On va réorganiser de la facon suivante
[  # l'identifiant de l'article est l'url
    #on le transfrome en id
    u'article_link',
    #date
    u'created_at',
    u'updated_at',
    #stats
    u'arguments_count',
    u'versions_count',
    u'votes_total',
    u'sources_count',
    
    # repartition des votes
    u'votes_ok',
    u'votes_mitige'
    u'votes_nok'
    #détails participation
    u'versions',
    u'sources',
    u'arguments',
    u'votes',
    ]

In [54]:
import json, pprint
ref_article9 = {"date": {}, "stats": {"votes":{}}, "details": {"votes":[]}, "votes":{}}
#pour afficher les données joliement
pp = pprint.PrettyPrinter(indent=4)
with open("article9.json", "r") as f:
    data = json.load(f)
    for k,v in data.items():
        
        if k == "article_link":
            ref_article9["id"] = v
        #match with date elemnts
        if k.endswith("at"):
            ref_article9["date"][k] = v
        elif k.endswith("count"):
            ref_article9["stats"][k.replace("count", "nb")] = v
        elif k.endswith("total"):
            ref_article9["stats"][k.replace("total", "nb")] = v
        elif k.startswith("votes"):
            ref_article9["stats"]["votes"][k.replace("votes_","")] = v
            
        else:
            if k in ["versions", "sources", "arguments"]:
                ref_article9["details"][k] = v
            pass
        #stats = {k:v for k,v in data.items() if k.endswith("at") and if  and if k.endswith("total") and if k.startswith("votes")}
pp.pprint(ref_article9["stats"])
#resultats des statistiques de l'article

{   u'arguments_nb': 132,
    u'sources_nb': 22,
    u'versions_nb': 108,
    'votes': {   u'mitige': 282, u'nok': 1732, u'ok': 1320},
    u'votes_nb': 3334}


L'article 9 compte donc 132 arguments,22 sources, 3334 votes et 108 versions.
Dans "details":
* Chaque argument contient le nombre de votes concernant l'argument
* Chaque argument contient le nombre de votes concernant la version


Le nombre de votes total correspond aux votes sur l'article original et le nombre d'arguments aussi
et non pas sur l'ensemble des versions de l'article


###Les modifications de l'article
Les amemendements ou modification de l'article sont disponible 
désormais dans ref_article9["details"]["versions"] sous forme de liste de dictionnaire
Explorons donc ensemble les différentes valeurs exploitables

In [17]:
versions = {}
for version in ref_article9["details"]["versions"]:
    print pp.pprint(version.keys())
    break

[   u'comment',
    u'author',
    u'arguments_no_count',
    u'created_at',
    u'after',
    u'updated_at',
    u'id',
    u'votes_mitige',
    u'arguments_yes_count',
    u'link',
    u'arguments_count',
    u'before',
    u'title',
    u'votes_total',
    u'article_id',
    u'slug',
    u'votes_nok']
None


Nous allons réorganiser le contenu des references de versions
pour rappatrier ensuite les votes qui contient les données détaillées 

[   #u'slug',
    #u'title'
    #date
    u'created_at',
    u'updated_at',
    u'author',
    #stats
        #arguments
        u'arguments_no_count',
        u'arguments_yes_count',
        u'arguments_count',
        #votes
        u'votes_total',
        u'votes_mitige',
        u'votes_nok',
        u'votes_ok'
    ]
dans un dictionnaire
dont l'id unique sera le slug
versions[slug] = {"title":"", "date":{}, "author": "", "stats":{"votes":{}, "arguments":{}}



In [65]:
versions = {}
for vers in ref_article9["details"]["versions"]:
    
    versions[vers["slug"]] = {"title":vers["title"], 
                              "date":{}, 
                              "author": vers["author"], 
                              "stats":{"arguments":{},"votes":{}},
                              "arguments":[],
                              "votes":[]
                             }
    for k,v in vers.items():
        
        if k.endswith("at"):
            versions[vers["slug"]]["date"][k] = v
        elif k.startswith("votes"):
            versions[vers["slug"]]["stats"]["votes"][k] = v
        elif k.startswith("arguments"):
            versions[vers["slug"]]["stats"]["arguments"][k] = v
    
#On a bien 108 versions dans ce dictionnaire
print len(versions.keys())
print versions["attention-a-la-qualite-et-a-la-securite-des-textes-scientifiques-qui-doivent-etre-valides-en-amont-promus-et-enrichis-ensuite-par-le-biais-de-bases-de-donnees"]

108
{'votes': [], 'stats': {'votes': {u'votes_mitige': 0, u'votes_total': 7, u'votes_nok': 3}, 'arguments': {u'arguments_yes_count': 0, u'arguments_no_count': 0, u'arguments_count': 0}}, 'author': u'dominiqueillien', 'title': u'Cr\xe9ation du nouvel article L 533-4 du code de la recherche (Livre V, Titre III, Chapitre 3)', 'arguments': [], 'date': {u'created_at': u'2015-10-12T16:46:41+0200', u'updated_at': u'2015-10-12T16:46:41+0200'}}


In [66]:
arguments = {}
for arg in ref_article9["details"]["arguments"]:
    #ici l'id unique est l'id
    #"arg-"+id
    arguments[arg["id"]] = {"date":{}, "stats":{"votes":{}}}
    for k, v in arg.items():
        if k.endswith("at"):
            arguments[arg["id"]]["date"][k] = v
    arguments[arg["id"]]["stats"]["votes"]["total"] = v
            

#on a bien les 132 arguments annoncés pour la version original 
print(len(arguments))

print arguments["3923"]

132
{'date': {u'created_at': u'2015-10-14T15:51:34+0200', u'updated_at': u'2015-10-14T15:52:07+0200'}, 'stats': {'votes': {'total': u'3923'}}}


Dans un deuxième temps nous allons croiser ces données avec ceux des votes
pour trouver qui a voté pour quelle modif ou quelle version
Pour cela il nous faut trouver tous les votes pour l'article9
Le détail des votes est disponible dans le fichier participants
qui liste sous forme de dictionnaire de participants toutes l'activité du participant


### Qui a voté quoi?
On va passer en revue tous les votes des participants et trouver ceux qui correspondent à l'article qui nous intéresse
en filtrant sur le nom de l'article contenu dans le lien
et construire ainsi un référentiel de votes


In [76]:
url = "section-2-travaux-de-recherche-et-de-statistique"
art = "article-9-acces-aux-travaux-de-la-recherche-financee-par-des-fonds-publics"
article9 = ref_article9
#raccourci
#votes = article9["details"]["votes"] = []
#arguments = article9["details"]["arguments"]
#version = article9["details"]["versions"]

with open("participants.json", "r") as f:
    data = json.load(f)
    participants = data["participants"]

#rappel versions contient toutes les versions de l'article
#on lui rappatrie les votes
#versions= []
ref_article9["details"]["votes"] = []
for participation in participants:
    for user, part in participation.items():
        for vote in part["votes"]:
            url = vote["link"]
            
            if art in url:
                
                votant = {"user":user, "date": vote["date"], "opinion": vote["opinion"], "id":vote["id"], "url":url}
                if "version" in url:
                    slug = url.split("/")[-1]
                    
                    if "arg" in url:
                        #ici il s'agit d'un vote sur l'argument de cette version
                        arg_id = slug.split("#arg-")[-1]
                        #versions[slug]["arguments"].append(votant)
                        #print arg_id
                    elif "source" in url:
                        pass
                    else:
                        versions[slug]["votes"].append(votant)
                else:
                    if "#arg" in url:
                        #un vote sur l'argument
                        pass
                    elif "#source" in url:
                        pass
                    else:
                        #ici il s'agit d'un vote sur l'article original
                        ref_article9["details"]["votes"].append(votant)
#on vérifie les votes
#votants
print len(ref_article9["details"]["votes"])
#votes comptabilisés dans l'article
print ref_article9["stats"]["votes_nb"]


3334
3334


In [ ]:
### Qui a voté pour qui? Pour quelle version?


In [86]:
#auteur, votant, vote, slug
final_dataset = []
for slug, version in versions.items():
    
    auteur = version["author"]
    for vote in version["votes"]:
        final_dataset.append([vote["user"],auteur,vote["opinion"], slug])
        

#On ajoute les votants à l'article original
for votant in ref_article9["details"]["votes"]:
    final_dataset.append([ vote["user"],"gouvernement",vote["opinion"], "original"])
    


In [92]:
with open("electeur_auteur.csv", "w") as f:
    f.write("\t".join(["electeur","auteur", "vote", "version"])+'\n')
    for row in final_dataset:
        f.write("\t".join(row)+'\n')
        
    

## Détection des communautés


In [ ]:
Nous avons désormais un fichier qui contient
l'électeur l'auteur le vote (-1 non 0 neutre 1 positif)

Nous avons réalisé dans un premier temps
une cartographie des coocurrences pas très réussie
